### ISSUE TITLE
About the method to calculate the global motion prior and local motion prior

### ISSUE THREAD

**Issue by bugsuse** [DeminYu98/DiffCast#1]
It's an interesting work! 

I found that you only use the last frame of the deterministic prediction $\mu$ when calculating the global motion prior, and the 1st frame of the deterministic prediction $\mu$  to calculate the local motion prior (it seems not in the paper?).

I am curious if the global motion prior information can be obtained using only the last frame of the deterministic prediction $\mu$? Looking forward to your reply! Thanks a lot!

https://github.com/DeminYu98/DiffCast/blob/7a08ad3ce2896691a95bfc819acda244620b8104/diffcast.py#L436-L447

**Comment by DeminYu98** [comment]
Nice question! Thank you for pointing it out.

Actually, the motion prior can be roughly regarded as the hidden state of ConvGRU, which records the historical temporal information from previous frames. Thus, the global motion prior is not calculated based on the last frame alone. 
However, the motion prior similarly faces the problem of long-term forgetting. It retains less information about the previous $\mu$, occasionally leading to the mismatch between $\mu$ and $r$ in the initial forecast, just as shown in following figure. 
Therefore, we manually add the first `local_ctx`. We did not mention local motion prior in our paper before we observed this problem.

Thanks again for your attention to our work.

![mismatch](https://github.com/DeminYu98/DiffCast/assets/88766469/de528ae7-bcde-49fb-b865-cc0294a09ead)

**Comment by bugsuse** [comment]
Thanks for your clarification and detailed explanation! 

I also observed the mismatch between the prediction and the ground truth when training the DiffCast using `SimVP` as the backbone for precipitation nowcasting (using 5 frames to predict 15 frames, splitting 3 segments). In addition, I encountered an issue with the discontinuity of the predictions of DiffCast at the sequence segment, as shown in the following figure.

Have you encountered this issue before? Why does this issue occur?


![valid_epoch6_batch400_index7](https://github.com/DeminYu98/DiffCast/assets/16102453/e0d2e04e-2247-4fce-8c8e-cf48a16583ab)

**Comment by DeminYu98** [comment]
As for mismatch, we only observe it in the fisrt forecast segment when making longer predictions. In practice, there should not be any mismatch at 5-->15 task, because the global motion prior can provide sufficient information during initial forecast.

As for discontinutiy, there are obvious unconsistancy at the 5th-6th and 10th-11th frames. However, we have not encountered this issue before.

Here are some suggestions you can consider:
- Calculate `diff_loss` among 3 segments and do backward at once following Fig 2(b) of our paper. If you have already done so, check for any issues with the **segment-contioned** UNet.
- Try to ignore the `local_ctx` or just set it to zeros.

**Comment by bugsuse** [comment]
Thanks for your suggestions. I will try it!

**Comment by bugsuse** [comment]
Sorry for disturbing you again. I have tested it based on your suggestions.

  
### Suggestion 1
> Calculate diff_loss among 3 segments and do backward at once following Fig 2(b) of our paper. If you have already done so, check for any issues with the segment-contioned UNet.

I am currently calculating the `diff_loss` among 3 segments, and after summing it with the loss of the deterministic backbone model, then do backward with the default `segment-contioned` UNet. The code is as follows:
```python
    def predict(self, frames_in, compute_loss=False, frames_gt=None, **kwargs):
        T_out = default(kwargs.get('T_out'), 15)

        if compute_loss:
            B, T_in, c, h, w = frames_in.shape
            device = self.device

            backbone_output, backbone_loss = self.backbone_net.predict(frames_in, frames_gt=frames_gt,
                                                                       compute_loss=compute_loss, **kwargs)

            residual = frames_gt - backbone_output
#            frames_in = self.normalize(frames_in)
#            frames_gt = self.normalize(frames_gt)
#            backbone_output = self.normalize(backbone_output)

            global_ctx, local_ctx = self.ctx_net.scan_ctx(torch.cat((frames_in, backbone_output), dim=1))

            pre_frag = frames_in
            pre_mu = None
            pred_ress = []
            diff_loss = 0.
            t = torch.randint(0, self.num_timesteps, (B,), device=self.device).long()
            for frag_idx in range(T_out // T_in):
                mu = backbone_output[:, frag_idx * T_in : (frag_idx + 1) * T_in]
                res = residual[:, frag_idx * T_in : (frag_idx + 1) * T_in]

                cond = pre_frag - pre_mu if pre_mu is not None else torch.zeros_like(pre_frag)
                res_pred, noise_loss = self.p_losses(res, t, cond=cond, ctx=global_ctx if frag_idx > 0 else local_ctx,
                                                     idx=torch.full((B,), frag_idx, device=device, dtype=torch.long))
                diff_loss += noise_loss

                frag_pred = res_pred + mu
                pre_frag = frag_pred
                pre_mu = mu

            alpha = torch.tensor(0.5)
            loss = (1 - alpha) * backbone_loss + alpha * diff_loss / 3.

            #backbone_output = self.unnormalize(backbone_output)
            return backbone_output, loss, backbone_loss, diff_loss
        else:
            pred, mu, y = self.sample(frames_in=frames_in, T_out=T_out)
            loss = None
            backbone_loss = None
            diff_loss = None

            return pred, mu, y, loss, backbone_loss, diff_loss
```
        
`p_loss` is also reference [denoising_diffusion_pytorch](https://github.com/lucidrains/denoising-diffusion-pytorch/blob/main/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py), the code as follows, and the objective is `pred_v`.
```python
    def p_losses(self, x_start, t, noise=None, offset_noise_strength=None, ctx=None, idx=None, cond=None):
        b, _, c, h, w = x_start.shape

        noise = default(noise, lambda: torch.randn_like(x_start))

        # offset noise - https://www.crosslabs.org/blog/diffusion-with-offset-noise
        offset_noise_strength = default(offset_noise_strength, self.offset_noise_strength)

        if offset_noise_strength > 0.:
            offset_noise = torch.randn(x_start.shape[:2], device = self.device)
            noise += offset_noise_strength * rearrange(offset_noise, 'b c -> b c 1 1')

        # noise sample
        x = self.q_sample(x_start=x_start, t=t, noise=noise)
        model_out = self.model(x, t, cond=cond, ctx=ctx, idx=idx)

        if self.objective == 'pred_noise':
            target = noise
        elif self.objective == 'pred_x0':
            target = x_start
        elif self.objective == 'pred_v':
            v = self.predict_v(x_start, t, noise)
            target = v
        else:
            raise ValueError(f'unknown objective {self.objective}')

        loss = F.mse_loss(model_out, target, reduction = 'none')
        loss = reduce(loss, 'b ... -> b', 'mean')

        loss = loss * extract(self.loss_weight, t, loss.shape)
        return model_out, loss.mean()
```  

 

### Suggestion 2
> Try to ignore the local_ctx or just set it to zeros.

I attempted to set `local_ctx` to zeros using the following code, but the prediction became much stranger than before, as shown in the subsequent figure. The residuals learned by the diffusion model appear to be entirely incorrect and do not correspond with the actual residuals.

```python
    def scan_ctx(self, frames):
        b, t, c, h, w = frames.shape
        state_shape = (b, c, h, w)
        self.init_state(state_shape, frames.device)
        local_ctx = None
        globla_ctx = None

        for i in range(t):
            globla_ctx = self.forward(frames[:,i])
            if i == 5:
                local_ctx = [torch.zeros_like(h.clone()) for h in globla_ctx]
        return globla_ctx, local_ctx
```

![valid_epoch5_batch0_index0](https://github.com/DeminYu98/DiffCast/assets/16102453/514d9b24-4fab-4763-8510-2b9a9de58579)

Do you have any suggestions to fix this problem? Any suggestions would be greatly appreciated!

**Comment by DeminYu98** [comment]
Sorry for the late reply 🥲.
The 'p_losses' func should not return `model_out` as the `res_pred`, which is meaningless.
We only aim to make `unet` learn to predict suitable nosie given sufficient condition, such as `previous_residual`, `global_context`, `frame_idx`, et al.
I think you can have a systematic understanding of diffusion models from the Training & Sampling Algorithm in DDPM [paper](https://arxiv.org/abs/2006.11239). 
Finally, you can determine if the diffusion model is being properly trained by observing the trend of the `diff_loss` during the training process.

**Comment by bugsuse** [comment]
Thanks for your reply.

How to get the `previous_residual` as the condition of `unet` model if not return `model_out` as the `res_pred` in the `p_losses` func?

Do you refer to the condition of `unet` model based on the following codes from `denoising-diffusion-pytorch`?

```python
        x_self_cond = None
        if self.self_condition and random() < 0.5:
            with torch.no_grad():
                x_self_cond = self.model_predictions(x, t).pred_x_start
                x_self_cond.detach_()
```

**Comment by DeminYu98** [comment]
Maybe there are some misunderstanding for our paper.
Since the diffusion aims to fit the residual distribution, we can get the overall residual sequence base on `mu` and `gt` during training, aka the following lines in your original code.
``` python
                residual = frames_gt - backbone_output
                res = residual[:, frag_idx * T_in : (frag_idx + 1) * T_in]
```
The Figure 2(b) in our paper shows the computing flow for training.
![WX20240510-114406@2x](https://github.com/DeminYu98/DiffCast/assets/88766469/735c930d-f6c0-43e1-8eaf-aeccd52b7b12)

The code of `denoising-diffusion-pytorch` in your reply is not necessary.

**Comment by bugsuse** [comment]
Thanks for your explanation.

I want to know how to get the condition as the `unet` model if not return `model_out` as the `res_pred`. 

Using the code you provided to calculate `residual`, the `res` should be used as the input of `self.model`, how to given the `cond` of the `self.model`?

The `pre_frag` should be the sum of `mu` and `res_pred`.  Returning `res_pred` is to calculate `residual_pred` as the condition of the `unet` model. How to get the `pre_frag` if not return `res_pred`?

```python
cond = pre_frag - pre_mu if pre_mu is not None else torch.zeros_like(pre_frag)
```

Do you calculate the `cond` of the `unet` model in the`p_losses` func using the above code? Or anything else? Or using the  mu as the condition of the `unet` model?

**Comment by DeminYu98** [comment]
Sorry, I donot understand your questions clearly.
1. the `unet` is one part of diffusion model, its output is just target noise, which can not be considered as final residual. We highly recommend the training and sampling algorithm of the DDPM paper.
2. the `pre_frag` is just the groud_truth of previous fragment.
3. `p_losses` func play the same role as origional DDPM with no extra calculation for residual condition.

Again, the figure  2b and our paper have explained the complete training procedure with diffusion training mechanism.

**Comment by bugsuse** [comment]
Thanks again for your detailed explanation. I have some misunderstandings about the training process. I will try it again!
